cumulusci/tasks/bulkdata/select_utils.py

```diff
@@ -0,0 +1,164 @@
+import typing as T
+
+from cumulusci.core.enums import StrEnum
+from cumulusci.tasks.bulkdata.extract_dataset_utils.hardcoded_default_declarations import (
+    DEFAULT_DECLARATIONS,
+)
+
+
+class SelectStrategy(StrEnum):
+    """Enum defining the different selection strategies requested."""
+
+    RANDOM = "random"
+    SIMILARITY = "similarity"
+
+
+def random_generate_query(
+    sobject: str, fields: T.List[str], num_records: float
+) -> T.Tuple[str, T.List[str]]:
+    """Generates the SOQL query for the random selection strategy"""
+    # Get the WHERE clause from DEFAULT_DECLARATIONS if available
+    declaration = DEFAULT_DECLARATIONS.get(sobject)
+    if declaration:
+        where_clause = declaration.where
+    else:
+        where_clause = None
+    # Construct the query with the WHERE clause (if it exists)
+    query = f"SELECT Id FROM {sobject}"
+    if where_clause:
+        query += f" WHERE {where_clause}"
+    query += f" LIMIT {num_records}"
+
+    return query, ["Id"]
+
+
+def random_post_process(
+    load_records, query_records: list, num_records: float, sobject: str
+) -> T.Tuple[T.List[dict], T.Union[str, None]]:
+    """Processes the query results for the random selection strategy"""
+    # Handle case where query returns 0 records
+    if not query_records:
+        error_message = f"No records found for {sobject} in the target org."
+        return [], error_message
+
+    # Add 'success: True' to each record to emulate records have been inserted
+    selected_records = [
+        {"id": record[0], "success": True, "created": False} for record in query_records
+    ]
+
+    # If fewer records than requested, repeat existing records to match num_records
+    if len(selected_records) < num_records:
+        original_records = selected_records.copy()
+        while len(selected_records) < num_records:
+            selected_records.extend(original_records)
+        selected_records = selected_records[:num_records]
+
+    return selected_records, None  # Return selected records and None for error
+
+
+def similarity_generate_query(
+    sobject: str,
+    fields: T.List[str],
+    num_records: float,
+) -> T.Tuple[str, T.List[str]]:
+    """Generates the SOQL query for the random selection strategy"""
+    # Get the WHERE clause from DEFAULT_DECLARATIONS if available
+    declaration = DEFAULT_DECLARATIONS.get(sobject)
+    if declaration:
+        where_clause = declaration.where
+    else:
+        where_clause = None
+    # Construct the query with the WHERE clause (if it exists)
+
+    fields.insert(0, "Id")
+    fields_to_query = ", ".join(field for field in fields if field)
+
+    query = f"SELECT {fields_to_query} FROM {sobject}"
+    if where_clause:
+        query += f" WHERE {where_clause}"
+
+    return query, fields
+
+
+def similarity_post_process(
+    load_records: list, query_records: list, num_records: float, sobject: str
```

Comment:
@mjawadtp remove the `list` typing for `load_records` as it is a generator

Reply:
I have removed `list` typing